common/sphinx.c

```diff
@@ -419,10 +419,12 @@ static void deserialize_hop_data(struct hop_data *data, const u8 *src)
 
 static void sphinx_write_frame(u8 *dest, const struct sphinx_hop *hop)
 {
-		memset(dest, 0, FRAME_SIZE);
-		dest[0] = hop->realm;
-		memcpy(dest + 1, hop->payload, tal_bytelen(hop->payload));
-		memcpy(dest + FRAME_SIZE - HMAC_SIZE, hop->hmac, HMAC_SIZE);
+	u8 num_frames = sphinx_hop_count_frames(hop);
+	memset(dest, 0, num_frames * FRAME_SIZE);
+	dest[0] = hop->realm | (num_frames-1) << 4;
```

Comment:
`|` binds less than `<<` but I still want more brackets here please :)  Also, this is now 3 is it not?

Maybe a heap of assertions here, like `assert(hop->realm < (1 << 4)` and `assert(tal_bytelen(hop->payload) < sphinx_hop_count_frames(hop) * FRAME_SIZE) - HMAC_SIZE - 1)`?